Expose an API to iterate all the registered operators

houseroad opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #23422 [WIP] Add CI job to check function schemas' backward compatiblity
* #23409 Add isBackwardCompatibleWith for Argument and FunctionSchema
* #23208 expose parse_schema and __eq__ function to python and add round trip tests
* **#23207 Expose an API to iterate all the registered operators**
* #23206 Align AliasInfo's operator<< with FunctionSchema

So we can iterate over the operator registry, and check the backward compatiblity.

Differential Revision: [D16434335](https://our.internmc.facebook.com/intern/diff/D16434335/)

houseroad committed:
Expose an API to iterate all the registered operators

So we can iterate over the operator registry, and check the backward compatiblity.

Differential Revision: [D16434335](https://our.internmc.facebook.com/intern/diff/D16434335/)

houseroad committed:
Update on "Expose an API to iterate all the registered operators"


So we can iterate over the operator registry, and check the backward compatiblity.

Differential Revision: [D16434335](https://our.internmc.facebook.com/intern/diff/D16434335/)

houseroad committed:
Update on "Expose an API to iterate all the registered operators"


So we can iterate over the operator registry, and check the backward compatiblity.

Differential Revision: [D16434335](https://our.internmc.facebook.com/intern/diff/D16434335/)

houseroad committed:
Update on "Expose an API to iterate all the registered operators"


So we can iterate over the operator registry, and check the backward compatiblity.

Differential Revision: [D16434335](https://our.internmc.facebook.com/intern/diff/D16434335/)

dzhulgakov commented on torch/csrc/jit/operator.cpp:
```diff
@@ -114,6 +114,17 @@ struct OperatorRegistry {
+  const std::vector<std::shared_ptr<Operator>>& getAllOperators() {
+    std::lock_guard<std::mutex> guard(lock);
+    registerPendingOperators();
+    static std::vector<std::shared_ptr<Operator>> values;
```
returning pointer to a static is not a good idea (as it can theoretically race. Since perf doesn't matter - just allocate new vector eac time

houseroad replied:
interesting... I thought with lock_guard, race should be gone.

Anyway, changed to allocating a new vector.

dzhulgakov replied:
you return the pointer to `values`, someone can read it when `guard` is no longer held (but instead if taken by another thread that repopulates `values` from scratch)

houseroad committed:
Update on "Expose an API to iterate all the registered operators"


So we can iterate over the operator registry, and check the backward compatiblity.

Differential Revision: [D16434335](https://our.internmc.facebook.com/intern/diff/D16434335/)

dzhulgakov commented on torch/csrc/jit/operator.cpp:
```diff
@@ -114,6 +114,17 @@ struct OperatorRegistry {
+    registerPendingOperators();
+    std::vector<std::shared_ptr<Operator>> values;
+    values.clear();
+    for (auto it = operators.begin(); it != operators.end(); ++it) {
```
make clang-tidy happy please

houseroad committed:
Update on "Expose an API to iterate all the registered operators"


So we can iterate over the operator registry, and check the backward compatiblity.

Differential Revision: [D16434335](https://our.internmc.facebook.com/intern/diff/D16434335/)

houseroad committed:
Update on "Expose an API to iterate all the registered operators"


So we can iterate over the operator registry, and check the backward compatiblity.

Differential Revision: [D16434335](https://our.internmc.facebook.com/intern/diff/D16434335/)

houseroad committed:
Update on "Expose an API to iterate all the registered operators"


So we can iterate over the operator registry, and check the backward compatiblity.

Differential Revision: [D16434335](https://our.internmc.facebook.com/intern/diff/D16434335/)
